Version upgrade guide (#288)
saleor/saleor-docs: docs/developer/upgrade-guide/2-11-to-3-0.mdx

---
title: Saleor 3.0
---

## Preparation

> New version introduces breaking changes and many new features. API clients like Apps and storefronts will require changes. 

Reviewer: ```suggestion
The new version introduces breaking changes and many new features. API clients like external apps and storefronts will require changes. 
```

> New version introduces breaking changes and many new features. API clients like Apps and storefronts will require changes. 

Reviewer 2: ```suggestion
The new version introduces many new features as well as some breaking changes. As a result, most API clients such as Saleor Apps and storefronts will require changes.
```

> Before you start the process, create the backup of your data. Also let your Celery worker to process task queue before starting API migration.

Reviewer: ```suggestion
Before you start the process, create a backup of your data. Also, let your Celery worker process the task queue before starting API migration.
```

> Due to scope of the upgrade, downtime will be required.

Reviewer: ```suggestion
Due to the scope of the upgrade, downtime will be required.
```

## Multichannel

> Existing data will be assigned to the `"default-channel"` created during the database migration. As long, as shop has the only one channel defined, you are not required to specify the `channel` argument. Thanks to that you'll be able to migrate storefront to 3.0 granually.

Reviewer: ```suggestion
Existing data will be assigned to the `"default-channel"` created during the database migration. As long, as the shop has only one channel defined, you are not required to specify the `channel` argument. Thanks to that you'll be able to migrate the storefront to 3.0 gradually.
```

> After adding more channels, argument will become required. Because of that, we recommend always adding the channel argument, even if you have only one channel:

Reviewer: ```suggestion
After adding more channels, the `channel` argument will become required. Because of that, we recommend always adding the channel argument, even if you have only one channel:
```

```graphql
{
  products(first: 5, channel: "default-channel") {
    edges {
      node {
        id
        name
        description
      }
    }
  }
}
```

Besides impact on fetching the data, you'll may be required to specify channel during creating the checkouts. You'll find updated guide at [Checkout](developer/checkout.mdx) page.

To read more about multichannel, check out [developer](developer/channels.mdx) and [dashboard](dashboard/channels.md) docs.

### Editing channel related fields

> If you use API to modify objects (for example with custom App), please note that channel depending fields have dedicated mutations: updating product price require using [productChannelListingUpdate](developer/api-reference/mutations/product-channel-listing-update.mdx) mutation.

Reviewer: ```suggestion
If you use API to modify objects (for example with external app), please note that channel depending fields have dedicated mutations: updating product price requires using [productChannelListingUpdate](developer/api-reference/mutations/product-channel-listing-update.mdx) mutation.
```

Object attributes which can be defined on channel basis can be reference
- [Collection Channel Listing](developer/api-reference/objects/collection-channel-listing.mdx)
- [Product Channel Listing](developer/api-reference/objects/product-channel-listing.mdx)
- [ProductVariant Channel Listing](developer/api-reference/objects/product-variant-channel-listing.mdx)
- [Sale Channel Listing](developer/api-reference/objects/sale-channel-listing.mdx)
- [ShippingMethod Channel Listing](developer/api-reference/objects/shipping-method-channel-listing.mdx)
- [Voucher Channel Listing](developer/api-reference/objects/voucher-channel-listing.mdx)

## Email templates

> New emails implementation brings easier configuration at dashboard and support for Sendgrid. If you modified default emails, you will need to migrate them to the Handlebars templates.

Reviewer: ```suggestion
In the new email implementation, the email templates are stored in the database instead of static server files. This change enables editing the templates in the dashboard and adds the integration with [Sendgrid](https://sendgrid.com/). If you modified the default email templates provided by Saleor, you need to migrate them to the [Handlebars](https://handlebarsjs.com/) format and configure them in the dashboard.
```

Read documentation about user and admin email plugins:
- [User emails](developer/available-plugins/user-emails.mdx)
- [Admin emails plugin](developer/available-plugins/admin-emails.mdx)

## Editor.js rich text editor and description fields

> All rich text fields now use Editor.js instead of Draft.js. Our database migration should convert old format to the new one, but visual inspection is recommended.

Reviewer: ```suggestion
All rich text fields now use [Editor.js](https://editorjs.io/) instead of [Draft.js](https://draftjs.org/). Our database migration should convert the old format to the new one, but visual inspection is recommended.
```

```graphql
query FeaturedProducts {
  collection(slug: "featured", channel: "default-channel") {
    products(first: 10) {
      edges {
        node {
          name
        }
      }
    }
  }
}
```

`Shop.navigation` has been dropped in favor of querying desired menu by slug:
```graphql
query MainMenu {
  menu(slug: "main-menu", channel: "default-channel") {
    name
    items {
      name
    }
  }
}
```

`Shop.geolocalization` has been removed.

## Error handling
The new version unifies error handling across the API. All mutation responses have an `errors` field, which contains issues information. For example:

```graphql
mutation {
  checkoutCreate(input: { lines: [] }) {
    errors {
      field
      message
      code
    }
  }
}
```

Server response:
```json
{
  "data": {
    "checkoutCreate": {
      "errors": [
        {
          "field": "channel",
          "message": "You need to provide channel slug.",
          "code": "MISSING_CHANNEL_SLUG"
        }
      ]
    }
  }
}
```

Error codes can be used for indentifing the issue and displaying custom message for the customers.

## Product filters

> API for searching and filtering the Products has changed.

Reviewer 3: ```suggestion
The API for searching and filtering products has been changed.
```

> Read documentation for the [product filters and sorting](developer/products.mdx#retrieving-a-list-of-products).

Reviewer 3: ```suggestion
Read documentation for [product filters and sorting](developer/products.mdx#retrieving-a-list-of-products).
```

## Attributes

`Attribute`:
- `Attribute.values` dropped in favor of paginated list called [`choices`](developer/attributes.mdx#fetching-all-attributes).
AttributeValue type:
- `Attribute.type: AttributeValueType` dropped. Use the `inputType` field to determine the type of attribute's value.

`AttributeValueInput`:
- `value: String` changed to `values` with type `[String!]`

> ## Invoice

Reviewer: ```suggestion
## Invoices
```

- invoiceSendEmail renamed to invoiceSendNotification

## Permissions

- `MANAGE_SERVICE_ACCOUNTS` removed. Use `MANAGE_APPS` for managing external apps.
- `MANAGE_PRODUCT_TYPES_AND_ATTRIBUTES` added. Staff users managing the accounts may need adding this permission
- `MANAGE_PAGE_TYPES_AND_ATTRIBUTES` added. Staff users managing pages may need adding this permission

## Apps

Service Accounts have been renamed to Apps.

> ### AuthorizationKey dropped

Reviewer: ```suggestion
### Authorization keys
```

> Dropped mutations

Reviewer: ```suggestion
The following mutations were dropped:
```
